### PR TITLE
PP-15572 Reject Adyen webhook with invalid HMAC signature

### PR DESCRIPTION
**Pull request by joshuapook-MT**
- Added exception handling and logging for errors while validating
- Added tests to check that invalid Hmac scenarios are handled as expected

Co-authored by @AbdirizakIdris @joshuapook-MT

### CODE REVIEW

File: src/test/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenRecurringTokenNotificationServiceTest.java [hunk 1]
```diff
@@ -118,6 +119,30 @@ void shouldRejectInvalidTokenNotification() {
...
-        verify(mockAppender, never()).doAppend(loggingEventArgumentCaptor.capture());
+        verify(mockAppender, atLeastOnce()).doAppend(loggingEventArgumentCaptor.capture());
+        List<LoggingEvent> loggingEvents = loggingEventArgumentCaptor.getAllValues();
+        assertThat(loggingEvents.stream()
+                        .anyMatch(event -> event.getFormattedMessage().equals("Hmac signature is invalid, rejecting Adyen token notification")),
+                is(true));
+    }
+
+    @Test
+    void shouldRejectITokenNotificationWhenExceptionIsThrown() {
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenRecurringTokenNotificationServiceTest.java]:
```suggestion
    void shouldRejectTokenNotificationWhenExceptionIsThrown() {
```

File: src/test/java/uk/gov/pay/connector/it/resources/adyen/AdyenTokensNotificationResourceIT.java [hunk 2]
```diff
@@ -60,6 +65,40 @@ void shouldRejectNotificationFromNonApprovedDomainForRecurringTokenNotification(
...
+                .contentType(APPLICATION_JSON)
+                .post(NOTIFICATION_PATH)
+                .then()
+                .statusCode(403);
+
+        logs.assertContains("Hmac signature is invalid, rejecting Adyen token notification");
+    }
+
+    @Test
+    void shouldRejectNotificationWithMissingHmacSignatureForRecurringTokenNotification() {
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/it/resources/adyen/AdyenTokensNotificationResourceIT.java]: Could we limit this to a single unhappy path, given that unit tests already cover all cases? Integration tests should be kept to a minimum because they are slow and time-consuming to execute.

Comment (joshuapook-MT) [reply src/test/java/uk/gov/pay/connector/it/resources/adyen/AdyenTokensNotificationResourceIT.java]: Deleted this test for the missing signature then, leaving the invalid one.

File: src/main/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenRecurringTokenNotificationService.java [hunk 3]
```diff
@@ -23,24 +24,40 @@ public AdyenRecurringTokenNotificationService(AdyenGatewayConfig adyenGatewayCon
...
 
     public boolean handleNotificationFor(String payload, String hmacSignature, String forwardedIpAddresses) {
-
-        if (!adyenNotificationValidator.isValidIpAddress(forwardedIpAddresses)) {
+        try {
+            if (!adyenNotificationValidator.isValidIpAddress(forwardedIpAddresses)) {
+                return false;
+            }
+
+            if (hmacSignature.isBlank()) {
```
Comment (Rafiq-Gasmi) [inline src/main/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenRecurringTokenNotificationService.java]: NPE if header is absent (null)?

File: src/main/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenRecurringTokenNotificationService.java [hunk 4]
```diff
@@ -23,24 +24,40 @@ public AdyenRecurringTokenNotificationService(AdyenGatewayConfig adyenGatewayCon
...
+                LOGGER.atInfo()
+                        .setMessage("Hmac signature is missing, rejecting Adyen token notification")
+                        .addKeyValue(PROVIDER, ADYEN.getName())
+                        .log();
+                return false;
+            }
+
+            // Todo: need to work out how to set this. Hardcoding for now
```
Comment (Rafiq-Gasmi) [inline src/main/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenRecurringTokenNotificationService.java:40]: Hardcoded live = false with unresolved TODO

File: src/test/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenNotificationValidatorTest.java
Comment (Rafiq-Gasmi) [inline src/test/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenNotificationValidatorTest.java:57]: Duplicate ArgumentCaptor fields

Comment (Rafiq-Gasmi) [inline src/test/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenNotificationValidatorTest.java]: Duplicate ArgumentCaptor fields

**Review by kbottla — DISMISSED** [review]
LGTM!

File: src/test/java/uk/gov/pay/connector/it/resources/adyen/AdyenTokensNotificationResourceIT.java [hunk 5]
```diff
@@ -1,9 +1,11 @@
 package uk.gov.pay.connector.it.resources.adyen;
```
Comment (Rafiq-Gasmi) [inline src/test/java/uk/gov/pay/connector/it/resources/adyen/AdyenTokensNotificationResourceIT.java:2]: Could add IT test for Scenario 3 (missing HMAC header)

File: src/test/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenRecurringTokenNotificationServiceTest.java
Comment (Rafiq-Gasmi) [inline src/test/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenRecurringTokenNotificationServiceTest.java]: AdyenNotificationValidator is fully mocked in this test so IpDomainMatcher is never really used. Remove to reduce noise.

File: src/test/java/uk/gov/pay/connector/it/resources/adyen/AdyenTokensNotificationResourceIT.java (hunk 5 above)
Comment (joshuapook-MT) [reply src/test/java/uk/gov/pay/connector/it/resources/adyen/AdyenTokensNotificationResourceIT.java:2]: I did have that but removed as Krishna pointed out we shouldn't have too many acceptance tests if we can help it (two very similar scenario's)

File: src/main/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenRecurringTokenNotificationService.java (hunk 4 above)
Comment (joshuapook-MT) [reply src/main/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenRecurringTokenNotificationService.java:40]: Yeah got the change for this coming in the next PR :)

File: src/test/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenNotificationValidatorTest.java
Comment (joshuapook-MT) [reply src/test/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenNotificationValidatorTest.java:57]: Fixed

File: src/main/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenRecurringTokenNotificationService.java (hunk 3 above)
Comment (joshuapook-MT) [reply src/main/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenRecurringTokenNotificationService.java]: Added null check

File: src/test/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenRecurringTokenNotificationServiceTest.java
Comment (joshuapook-MT) [reply src/test/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenRecurringTokenNotificationServiceTest.java]: There is a test that verify's it's not called below on line 75

Comment (Rafiq-Gasmi) [reply src/test/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenRecurringTokenNotificationServiceTest.java]: This isn't verifying anything meaningful as AdyenNotificationValidator is mocked and IpDomainMatcher is inside

Comment (joshuapook-MT) [reply src/test/java/uk/gov/pay/connector/gateway/adyen/webhook/AdyenRecurringTokenNotificationServiceTest.java]: Removed

**Review by Rafiq-Gasmi — APPROVED** [review]
LGTM
